File: examples/motoko_examples/quicksort/canisters/rust/src/quicksort.rs
use ic_cdk::export::candid::Int;
use std::{cmp::Ordering, convert::TryInto};

trait AbsoluteValue {
    fn abs(&self) -> usize;
}

impl AbsoluteValue for Int {
    fn abs(&self) -> usize {
        let positive_int = if self >= &Int::from(0i8) {
            self.clone()
        } else {
            self.clone() * Int::from(-1i8)
        };
        positive_int.0.try_into().unwrap()
    }
}

pub fn sort_by<X: Clone + Ord, T: Fn(&X, &X) -> Ordering>(xs: Vec<X>, f: &T) -> Vec<X> {
    let n = xs.len();
    if n < 2 {
        return xs;
    } else {
        let mut result = xs.clone();
        sort_by_helper(&mut result, 0.into(), (n - 1).into(), f);
        result
    }
}

fn sort_by_helper<X: Clone + Ord, T: Fn(&X, &X) -> Ordering>(
    xs: &mut Vec<X>,
    l: Int,
    r: Int,
    f: &T,
) -> () {
    if l < r {
        let mut i = l.clone();
        let mut j = r.clone();
        let pivot = xs[(l.clone() + r.clone()).abs() / 2].clone();
        while i <= j {
            while f(&xs[i.abs()], &pivot).is_lt() {
                i += 1;
            }
            while f(&xs[j.abs()], &pivot).is_gt() {
                j -= 1;
            }
            if i <= j {
                xs.swap(i.abs(), j.abs());
                i += 1;
                j -= 1;
            }
        }
        if l < j {
            sort_by_helper(xs, l, j, f);
        }
        if i < r {
            sort_by_helper(xs, i, r, f);
        }
    }
}

File: examples/motoko_examples/quicksort/src/quicksort.ts
import { Int, Order, OrderType } from './comparison';

export function sortBy<X>(xs: X[], f: (a: X, b: X) => OrderType): X[] {
    const n = xs.length;
    if (n < 2) {
        return xs;
    } else {
        let result = [...xs];
        sortByHelper(result, 0, n - 1, f);
        return result;
    }
}

function sortByHelper<X>(
    xs: X[],
    l: number,
    r: number,
    f: (a: X, b: X) => OrderType
): void {
    if (l < r) {
        let i = l;
        let j = r;
        let swap = xs[0];
        const pivot = xs[Math.round(Int.abs(l + r) / 2)];
        while (i <= j) {
            while (Order.isLess(f(xs[Int.abs(i)], pivot))) {
                i += 1;
            }
            while (Order.isGreater(f(xs[Int.abs(j)], pivot))) {
                j -= 1;
            }
            if (i <= j) {
                swap = xs[Int.abs(i)];
                xs[Int.abs(i)] = xs[Int.abs(j)];
                xs[Int.abs(j)] = swap;
                i += 1;
                j -= 1;
            }
        }
        if (l < j) {
            sortByHelper<X>(xs, l, j, f);
        }
        if (i < r) {
            sortByHelper<X>(xs, i, r, f);
        }
    }
}
